torchtext/datasets/text_classification.py

```diff
@@ -113,33 +119,61 @@ def get_vocab(self):
         return self._vocab
 
 
-def _setup_datasets(dataset_name, root='.data', ngrams=2, vocab=None, include_unk=False):
+def _setup_datasets(dataset_name, root='.data', ngrams=2, vocab=None,
+                    removed_tokens=[], tokenizer=None,
+                    data_select=('train', 'test')):
+    if not tokenizer:
+        tokenizer = get_tokenizer("basic_english")
+
+    if isinstance(data_select, str):
+        data_select = [data_select]
+    if not set(data_select).issubset(set(('train', 'test'))):
+        raise TypeError('data_select is not supported!')
+
     dataset_tar = download_from_url(URLS[dataset_name], root=root)
     extracted_files = extract_archive(dataset_tar)
 
+    path = {}
     for fname in extracted_files:
-        if fname.endswith('train.csv'):
-            train_csv_path = fname
-        if fname.endswith('test.csv'):
-            test_csv_path = fname
+        if fname.endswith('train.csv') and 'train' in data_select:
+            path['train'] = fname
+        if fname.endswith('test.csv') and 'test' in data_select:
+            path['test'] = fname
 
     if vocab is None:
-        logging.info('Building Vocab based on {}'.format(train_csv_path))
-        vocab = build_vocab_from_iterator(_csv_iterator(train_csv_path, ngrams))
+        if 'train' not in data_select:
+            raise TypeError("Must pass a vocab if train is not selected.")
+        logging.info('Building Vocab based on train data')
+        if dataset_name == 'IMDB':
+            data_iterator = _imdb_iterator('train', extracted_files, tokenizer, ngrams)
+        else:
+            data_iterator = _csv_iterator(path['train'], tokenizer, ngrams)
+        vocab = build_vocab_from_iterator(data_iterator)
     else:
         if not isinstance(vocab, Vocab):
             raise TypeError("Passed vocabulary is not of type Vocab")
     logging.info('Vocab has {} entries'.format(len(vocab)))
-    logging.info('Creating training data')
-    train_data, train_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(train_csv_path, ngrams, yield_cls=True), include_unk)
-    logging.info('Creating testing data')
-    test_data, test_labels = _create_data_from_iterator(
-        vocab, _csv_iterator(test_csv_path, ngrams, yield_cls=True), include_unk)
-    if len(train_labels ^ test_labels) > 0:
-        raise ValueError("Training and test labels don't match")
-    return (TextClassificationDataset(vocab, train_data, train_labels),
-            TextClassificationDataset(vocab, test_data, test_labels))
+
+    data = {}
+    for item in data_select:
+        data[item] = {}
+        data[item]['data'] = []
+        data[item]['labels'] = []
+        logging.info('Creating {} data'.format(item))
+        if dataset_name == 'IMDB':
+            data_iterator = _imdb_iterator(item, extracted_files,
```

Comment:
I understand that you need to generate the iterator twice, so we could also look into passing a function that is a generator factory to this setup function that'll allow you to generate an iterator over a particular dataset. That should keep this function shorter.